### ISSUE TITLE
`isinstance` narrowing keeps a non-matching union member as `X & Top[Generic[Unknown]]`, causing a false `invalid-argument-type`

### ISSUE THREAD

**Issue by adamtheturtle** [astral-sh/ty#3817]
Follow-up to #456, which is closed but still reproduces on ty 0.0.51.

After `isinstance(x, A)` on a union, ty recovers the specialized arm but keeps the non-matching member as an `X & Top[A[Unknown]]` intersection instead of dropping it.
That intersection is then rejected as not assignable to a concrete `A[...]` parameter.
This is a false positive: after the guard the value is an `A`.

## Reproduction

```python
from typing import Generic, TypeVar

T = TypeVar("T")

class A(Generic[T]): ...

def f(x: A[int]) -> None: ...

def g(x: A[int] | int) -> None:
    if isinstance(x, A):
        reveal_type(x)   # ty: A[int] | (int & Top[A[Unknown]]); expected A[int]
        f(x)             # ty: error[invalid-argument-type]
```

mypy 1.18.2 and pyright 1.1.407 both narrow the true branch to `A[int]` and accept `f(x)`.

**Comment by sharkdp** [comment]
Please see https://docs.astral.sh/ty/reference/typing-faq/#what-is-toplistunknown-and-why-does-it-appear. There could be a common subclass of `A[int]` and `int`. You can fix this by making `A` `@final`.

**Comment by carljm** [comment]
This is #1578.

> There could be a common subclass of `A[int]` and `int`

Nit: the issue is that there could be a common subclass of `A` and `int`, which is not an `A[int]`.
